# NEURO BAND — Morning Brief System
### Your AI-powered daily dashboard: Flashcards • To-Do • Weather • Health

---
**What this notebook does:**
- 📚 **Flashcard Tracker** — Record & review yesterday's study cards
- ✅ **To-Do List** — Manage daily tasks with priority levels
- 🌤️ **Weather** — Live weather via `wttr.in` (100% FREE, no key needed)
- 🧬 **Health AI** — Personalized health tips via HuggingFace FREE API
- 🖥️ **Morning Brief** — Beautiful all-in-one dashboard

**Run cells in order: Cell 1 → 2 → 3 → 4 → 5 → 6 → 7 → 8**

In [ ]:
# ============================================================
# CELL 1 — INSTALL LIBRARIES
# Run this ONCE at the start of every Colab session
# ============================================================

!pip install requests ipywidgets -q

# Enable ipywidgets display in Colab
from google.colab import output
output.enable_custom_widget_manager()

print('✅ All libraries installed and ready!')
print('👉 Now run Cell 2 to configure your settings.')

✅ All libraries installed and ready!
👉 Now run Cell 2 to configure your settings.


In [ ]:
# ============================================================
# CELL 2 — CONFIGURATION
# Edit these 3 values before running
# ============================================================

# ✏️ YOUR SETTINGS — EDIT THESE:
YOUR_CITY      = "Colombo"          # e.g., "London", "Mumbai", "New York"
YOUR_NAME      = "Banu Athuraliya"           # Your name for the greeting
HF_TOKEN       = "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"  # From huggingface.co/settings/tokens (FREE)

# ─── HuggingFace FREE Model (best free text-gen model) ──────
HF_MODEL = "HuggingFaceH4/zephyr-7b-beta"  # Best free model — do NOT change
HF_API_URL = f"https://api-inference.huggingface.co/models/{HF_MODEL}"

# ─── Storage file (saved inside Colab /content/) ─────────────
import os
DATA_FILE = "/content/neuro_band_data.json"

print('✅ Configuration saved!')
print(f'   📍 City   : {YOUR_CITY}')
print(f'   👤 Name   : {YOUR_NAME}')
print(f'   🤖 Model  : {HF_MODEL}')
print()

print('👉 Now run Cell 3 to set up the data manager.')

✅ Configuration saved!
   📍 City   : Colombo
   👤 Name   : Friend
   🤖 Model  : HuggingFaceH4/zephyr-7b-beta

👉 Now run Cell 3 to set up the data manager.


In [ ]:
# ============================================================
# CELL 3 — DATA MANAGER (Runs silently in background)
# ============================================================

import json
import requests
from datetime import datetime, date, timedelta
from IPython.display import display, HTML, clear_output
import ipywidgets as widgets

# ─── Load / Save helpers ─────────────────────────────────────
def load_data():
    """Load all saved data from the JSON file."""
    default = {
        "flashcards": {},   # { "YYYY-MM-DD": [ {topic, question, answer, time}, ... ] }
        "todos": [],        # [ {task, priority, done, created, id}, ... ]
    }
    if os.path.exists(DATA_FILE):
        try:
            with open(DATA_FILE, 'r') as f:
                return json.load(f)
        except Exception:
            return default
    return default

def save_data(data):
    """Save all data to the JSON file."""
    with open(DATA_FILE, 'w') as f:
        json.dump(data, f, indent=2)

# ─── Initial load ─────────────────────────────────────────────
data = load_data()
today_str     = str(date.today())
yesterday_str = str(date.today() - timedelta(days=1))

print('✅ Data Manager Ready!')
print(f'   📂 Storage file : {DATA_FILE}')
print(f'   📚 Flashcard days on record : {len(data["flashcards"])}')
print(f'   📋 Total to-do items        : {len(data["todos"])}')
print()
print('👉 Now run Cell 4 — Flashcard Manager')

✅ Data Manager Ready!
   📂 Storage file : /content/neuro_band_data.json
   📚 Flashcard days on record : 0
   📋 Total to-do items        : 0

👉 Now run Cell 4 — Flashcard Manager


In [ ]:
# ============================================================
# CELL 4 — FLASHCARD MANAGER
# Add, view, and review your study flashcards
# ============================================================

def flashcard_manager():
    data = load_data()
    out  = widgets.Output()

    # ── Header ──────────────────────────────────────────────
    header = widgets.HTML(value="""
    <div style='background:linear-gradient(135deg,#667eea,#764ba2);
                padding:14px 18px; border-radius:10px; color:white; margin-bottom:12px;'>
        <h2 style='margin:0; font-family:sans-serif'>📚 Flashcard Manager</h2>
        <p  style='margin:4px 0 0; opacity:.75; font-family:sans-serif; font-size:.85em'>
            Add cards you studied today. They will appear in tomorrow's Morning Brief.
        </p>
    </div>
    """)

    # ── Input widgets ────────────────────────────────────────
    topic_in = widgets.Text(
        placeholder='Topic (e.g., Biology, Python)',
        layout=widgets.Layout(width='280px'))

    question_in = widgets.Textarea(
        placeholder='Question',
        layout=widgets.Layout(width='460px', height='60px'))

    answer_in = widgets.Textarea(
        placeholder='Answer',
        layout=widgets.Layout(width='460px', height='60px'))

    add_btn      = widgets.Button(description='➕ Add Card',         button_style='primary')
    view_btn     = widgets.Button(description='📖 View Today\'s',    button_style='info')
    view_yest_btn= widgets.Button(description='📅 View Yesterday\'s',button_style='warning')

    # ── Button callbacks ─────────────────────────────────────
    def add_card(b):
        data = load_data()
        today = str(date.today())
        q = question_in.value.strip()
        a = answer_in.value.strip()
        if not q or not a:
            with out:
                clear_output()
                print('⚠️  Please fill in both Question and Answer fields.')
            return
        if today not in data['flashcards']:
            data['flashcards'][today] = []
        data['flashcards'][today].append({
            'topic'    : topic_in.value.strip() or 'General',
            'question' : q,
            'answer'   : a,
            'time'     : datetime.now().strftime('%H:%M')
        })
        save_data(data)
        topic_in.value    = ''
        question_in.value = ''
        answer_in.value   = ''
        count = len(data['flashcards'][today])
        with out:
            clear_output()
            print(f'✅ Flashcard added! You now have {count} card(s) for today ({today}).')

    def view_today(b):
        data  = load_data()
        today = str(date.today())
        cards = data['flashcards'].get(today, [])
        with out:
            clear_output()
            if not cards:
                print("No flashcards added today yet. Use the form above to add some!")
                return
            print(f"📚 TODAY'S FLASHCARDS ({today}) — {len(cards)} total")
            print('='*50)
            for i, c in enumerate(cards, 1):
                print(f"\n  Card {i} | 🏷️ {c['topic']} | ⏰ {c['time']}")
                print(f"  ❓ Q: {c['question']}")
                print(f"  💡 A: {c['answer']}")

    def view_yesterday(b):
        data      = load_data()
        yesterday = str(date.today() - timedelta(days=1))
        cards     = data['flashcards'].get(yesterday, [])
        with out:
            clear_output()
            if not cards:
                print(f"No flashcards found for yesterday ({yesterday}).")
                return
            print(f"📅 YESTERDAY'S FLASHCARDS ({yesterday}) — {len(cards)} total")
            print('='*50)
            for i, c in enumerate(cards, 1):
                print(f"\n  Card {i} | 🏷️ {c['topic']} | ⏰ {c['time']}")
                print(f"  ❓ Q: {c['question']}")
                print(f"  💡 A: {c['answer']}")

    add_btn.on_click(add_card)
    view_btn.on_click(view_today)
    view_yest_btn.on_click(view_yesterday)

    display(
        header,
        widgets.HBox([topic_in]),
        question_in,
        answer_in,
        widgets.HBox([add_btn, view_btn, view_yest_btn],
                     layout=widgets.Layout(gap='8px', margin='8px 0')),
        out
    )

flashcard_manager()
print('\n👉 After adding cards, run Cell 5 — To-Do List')

HTML(value="\n    <div style='background:linear-gradient(135deg,#667eea,#764ba2);\n                padding:14p…

Textarea(value='', layout=Layout(height='60px', width='460px'), placeholder='Question')

Textarea(value='', layout=Layout(height='60px', width='460px'), placeholder='Answer')

Output()


👉 After adding cards, run Cell 5 — To-Do List


In [ ]:
# ============================================================
# CELL 5 — TO-DO LIST MANAGER
# Add tasks, set priorities, mark them done
# ============================================================

def todo_manager():
    data = load_data()
    out  = widgets.Output()

    # ── Header ──────────────────────────────────────────────
    header = widgets.HTML(value="""
    <div style='background:linear-gradient(135deg,#11998e,#38ef7d);
                padding:14px 18px; border-radius:10px; color:white; margin-bottom:12px;'>
        <h2 style='margin:0; font-family:sans-serif'>✅ To-Do List Manager</h2>
        <p  style='margin:4px 0 0; opacity:.8; font-family:sans-serif; font-size:.85em'>
            Add, view, complete, and clear your daily tasks.
        </p>
    </div>
    """)

    # ── Input widgets ────────────────────────────────────────
    task_in = widgets.Text(
        placeholder='Enter a task...',
        layout=widgets.Layout(width='350px'))

    priority_drop = widgets.Dropdown(
        options=['🔴 High', '🟡 Medium', '🟢 Low'],
        value='🟡 Medium',
        layout=widgets.Layout(width='140px'))

    add_btn   = widgets.Button(description='➕ Add Task',     button_style='primary')
    view_btn  = widgets.Button(description='📋 View All',     button_style='info')

    done_num  = widgets.BoundedIntText(
        min=0, max=999, value=0,
        description='Task #:',
        layout=widgets.Layout(width='160px'))

    done_btn  = widgets.Button(description='✓ Mark Done',     button_style='success')
    clear_btn = widgets.Button(description='🗑️ Clear Done',   button_style='warning')

    # ── Callbacks ────────────────────────────────────────────
    def add_task(b):
        data = load_data()
        task = task_in.value.strip()
        if not task:
            with out:
                clear_output()
                print('⚠️  Task cannot be empty.')
            return
        new_id = max([t.get('id', 0) for t in data['todos']], default=0) + 1
        data['todos'].append({
            'id'      : new_id,
            'task'    : task,
            'priority': priority_drop.value,
            'done'    : False,
            'created' : str(date.today())
        })
        save_data(data)
        task_in.value = ''
        with out:
            clear_output()
            pending = sum(1 for t in data['todos'] if not t['done'])
            print(f'✅ Task added! You have {pending} pending task(s).')

    def view_tasks(b):
        data    = load_data()
        pending = [t for t in data['todos'] if not t['done']]
        done    = [t for t in data['todos'] if  t['done']]
        with out:
            clear_output()
            print(f"\n📋 PENDING TASKS ({len(pending)})")
            print('='*50)
            if not pending:
                print('  ✨ No pending tasks! You are all caught up.')
            else:
                for i, t in enumerate(pending):
                    print(f"  [{i}] {t['priority']}  |  {t['task']}  (added: {t['created']})")
            print(f"\n✅ COMPLETED TASKS ({len(done)})")
            print('='*50)
            if not done:
                print('  None completed yet.')
            else:
                for t in done:
                    print(f"  ✓  {t['task']}")

    def mark_done(b):
        data    = load_data()
        pending = [t for t in data['todos'] if not t['done']]
        idx     = done_num.value
        if not pending:
            with out:
                clear_output()
                print('ℹ️  No pending tasks to complete.')
            return
        if idx < 0 or idx >= len(pending):
            with out:
                clear_output()
                print(f'⚠️  Invalid task number. Enter 0–{len(pending)-1}.')
                print('Tip: Click "📋 View All" first to see task numbers.')
            return
        # Mark it done in the master list
        task_name = pending[idx]['task']
        target_id = pending[idx]['id']
        for t in data['todos']:
            if t['id'] == target_id:
                t['done'] = True
                break
        save_data(data)
        with out:
            clear_output()
            print(f'✅  "{task_name}" marked as done! Great work!')

    def clear_done(b):
        data = load_data()
        removed = sum(1 for t in data['todos'] if t['done'])
        data['todos'] = [t for t in data['todos'] if not t['done']]
        save_data(data)
        with out:
            clear_output()
            print(f'🗑️  Cleared {removed} completed task(s). Clean slate!')

    add_btn.on_click(add_task)
    view_btn.on_click(view_tasks)
    done_btn.on_click(mark_done)
    clear_btn.on_click(clear_done)

    display(
        header,
        widgets.HBox([task_in, priority_drop, add_btn],
                     layout=widgets.Layout(gap='6px', align_items='center')),
        widgets.HBox([view_btn, clear_btn],
                     layout=widgets.Layout(gap='6px', margin='6px 0')),
        widgets.HTML(value="<p style='font-family:sans-serif; font-size:.85em; margin:4px 0; color:#555;'>"
                           "To mark a task done: click 📋 View All to see numbers, then enter the number below.</p>"),
        widgets.HBox([done_num, done_btn],
                     layout=widgets.Layout(gap='6px', align_items='center')),
        out
    )

todo_manager()
print('\n👉 After setting up your tasks, run Cell 6 — Weather')

HTML(value="\n    <div style='background:linear-gradient(135deg,#11998e,#38ef7d);\n                padding:14p…

HTML(value="<p style='font-family:sans-serif; font-size:.85em; margin:4px 0; color:#555;'>To mark a task done:…

Output()


👉 After setting up your tasks, run Cell 6 — Weather


In [ ]:
# ============================================================
# CELL 6 — LIVE WEATHER  (wttr.in — 100% FREE, NO API KEY)
# ============================================================

def get_weather(city=YOUR_CITY):
    """Fetch live weather from wttr.in (completely free, no API key required)."""
    print(f'🌐 Fetching weather for {city}...')
    try:
        url  = f'https://wttr.in/{city.replace(" ","+")}?format=j1'
        resp = requests.get(url, timeout=15)
        resp.raise_for_status()
        raw  = resp.json()

        cur  = raw['current_condition'][0]
        tod  = raw['weather'][0]

        info = {
            'city'       : city,
            'desc'       : cur['weatherDesc'][0]['value'],
            'temp_c'     : cur['temp_C'],
            'feels_c'    : cur['FeelsLikeC'],
            'humidity'   : cur['humidity'],
            'wind_kmph'  : cur['windspeedKmph'],
            'uv_index'   : cur['uvIndex'],
            'visibility' : cur['visibility'],
            'min_c'      : tod['mintempC'],
            'max_c'      : tod['maxtempC'],
        }

        print()
        print('🌤️  LIVE WEATHER REPORT')
        print('='*44)
        print(f"  📍 City        : {info['city']}")
        print(f"  🌡️  Temperature : {info['temp_c']}°C  (feels like {info['feels_c']}°C)")
        print(f"  ☁️  Condition   : {info['desc']}")
        print(f"  💧 Humidity    : {info['humidity']}%")
        print(f"  💨 Wind        : {info['wind_kmph']} km/h")
        print(f"  ☀️  UV Index    : {info['uv_index']}")
        print(f"  👁️  Visibility  : {info['visibility']} km")
        print(f"  📊 Today Range : {info['min_c']}°C — {info['max_c']}°C")
        print()
        print('✅ Weather fetched successfully!')
        return info

    except requests.exceptions.ConnectionError:
        print('❌ No internet connection. Check your network.')
    except requests.exceptions.Timeout:
        print('❌ Request timed out. Try again in a moment.')
    except Exception as e:
        print(f'❌ Weather fetch error: {e}')

    # Fallback so later cells don't crash
    return {
        'city':'Unknown','desc':'Unavailable','temp_c':'N/A','feels_c':'N/A',
        'humidity':'N/A','wind_kmph':'N/A','uv_index':'0',
        'visibility':'N/A','min_c':'N/A','max_c':'N/A'
    }


# Run it
weather_info = get_weather(YOUR_CITY)
print('\n👉 Now run Cell 7 — AI Health Recommendations')

🌐 Fetching weather for Colombo...

🌤️  LIVE WEATHER REPORT
  📍 City        : Colombo
  🌡️  Temperature : 27°C  (feels like 31°C)
  ☁️  Condition   : Partly cloudy
  💧 Humidity    : 94%
  💨 Wind        : 6 km/h
  ☀️  UV Index    : 3
  👁️  Visibility  : 7 km
  📊 Today Range : 25°C — 30°C

✅ Weather fetched successfully!

👉 Now run Cell 7 — AI Health Recommendations


In [ ]:
# ============================================================
# CELL 7 — AI HEALTH RECOMMENDATIONS (HuggingFace FREE API)
# Model: HuggingFaceH4/zephyr-7b-beta
# ============================================================

import time

def get_health_recommendations(weather, todos, yesterday_cards):
    """Call HuggingFace Inference API for personalised health tips."""

    pending_count  = sum(1 for t in todos if not t['done'])
    card_count     = len(yesterday_cards)
    topics_studied = list(set(c.get('topic', 'General') for c in yesterday_cards)) if yesterday_cards else []

    prompt = f"""<|system|>
You are a concise health & wellness coach. Give exactly 5 practical health tips.
Each tip must be one sentence. Number them 1 to 5. No extra text.<|endoftext|>
<|user|>
My morning data:
- Location: {weather.get('city','Unknown')}
- Weather: {weather.get('desc','Unknown')}, {weather.get('temp_c','?')}°C
- Humidity: {weather.get('humidity','?')}%
- UV Index: {weather.get('uv_index','?')}
- Wind: {weather.get('wind_kmph','?')} km/h
- Pending tasks today: {pending_count}
- Flashcards studied yesterday: {card_count} ({', '.join(topics_studied) if topics_studied else 'none'})

Give me 5 specific health tips for today based on this data.<|endoftext|>
<|assistant|>"""

    headers = {"Authorization": f"Bearer {HF_TOKEN}"}
    payload = {
        "inputs": prompt,
        "parameters": {
            "max_new_tokens": 350,
            "temperature": 0.7,
            "do_sample": True,
            "return_full_text": False,
            "stop": ["<|endoftext|>", "<|user|>"]
        }
    }

    print('🤖 Calling HuggingFace AI... (may take 15–45 seconds on first call)')
    print('   Model is loading if this is your first call today — please wait.')

    for attempt in range(3):  # Retry up to 3 times (model may be loading)
        try:
            resp = requests.post(HF_API_URL, headers=headers, json=payload, timeout=90)

            if resp.status_code == 503:  # Model loading
                wait = resp.json().get('estimated_time', 20)
                print(f'   ⏳ Model loading... waiting {int(wait)+5}s (attempt {attempt+1}/3)')
                time.sleep(int(wait) + 5)
                continue

            if resp.status_code == 401:
                print('❌ Invalid HuggingFace token. Check Cell 2 and re-run.')
                return None

            resp.raise_for_status()
            result = resp.json()

            if isinstance(result, list) and result:
                text = result[0].get('generated_text', '').strip()
                print()
                print('🧬 AI HEALTH RECOMMENDATIONS FOR TODAY')
                print('='*44)
                print(text)
                print()
                print('✅ AI recommendations generated!')
                return text
            else:
                raise ValueError(f'Unexpected API response: {result}')

        except requests.exceptions.Timeout:
            print(f'   ⏳ Timeout on attempt {attempt+1}. Retrying...')
            time.sleep(10)
        except Exception as e:
            print(f'   ⚠️  Attempt {attempt+1} failed: {e}')
            time.sleep(5)

    # ── Fallback: rule-based tips if API unavailable ─────────
    print()
    print('⚠️  HuggingFace API unavailable. Using smart fallback recommendations.')
    print()
    tips = []
    uv   = int(weather.get('uv_index', 0)) if str(weather.get('uv_index','0')).isdigit() else 0
    hum  = int(weather.get('humidity', 50)) if str(weather.get('humidity','50')).isdigit() else 50
    temp = int(weather.get('temp_c', 25)) if str(weather.get('temp_c','25')).lstrip('-').isdigit() else 25
    wind = int(weather.get('wind_kmph', 0)) if str(weather.get('wind_kmph','0')).isdigit() else 0

    if uv >= 8:
        tips.append('1. ☀️ UV index is very high — apply SPF 50+ sunscreen and wear a hat outdoors.')
    elif uv >= 5:
        tips.append('1. 🧴 UV index is moderate-high — use SPF 30 sunscreen if going outside.')
    else:
        tips.append('1. 🚶 UV index is low — great day for an outdoor walk or light exercise.')

    if hum > 75:
        tips.append('2. 💧 High humidity today — drink an extra 500ml of water and stay in shade.')
    elif hum < 35:
        tips.append('2. 💧 Low humidity — moisturise your skin and drink water every 45 minutes.')
    else:
        tips.append('2. 💧 Humidity is comfortable — aim for 8 glasses of water throughout the day.')

    if temp > 32:
        tips.append('3. 🧊 It is very hot — avoid outdoor activity between 11am–3pm; eat light meals.')
    elif temp < 18:
        tips.append('3. 🧥 It is cool today — layer up before heading out to protect your joints.')
    else:
        tips.append('3. 🌿 Great temperature for a 20-minute morning walk to boost your metabolism.')

    if pending_count > 5:
        tips.append('4. 🧠 You have many tasks — use the Pomodoro technique: 25 min work, 5 min break.')
    else:
        tips.append('4. 🧠 Follow the 20-20-20 rule: every 20 minutes look 20 feet away for 20 seconds.')

    if card_count > 0:
        tips.append(f'5. 📚 You studied {card_count} flashcard(s) yesterday — quick review before bed locks in memory.')
    else:
        tips.append('5. 📚 No flashcards yesterday — start small: add 5 cards today to build a study habit.')

    result_text = '\n'.join(tips)
    print('🧬 SMART HEALTH RECOMMENDATIONS FOR TODAY')
    print('='*44)
    print(result_text)
    return result_text


# ── Run it ───────────────────────────────────────────────────
data_now       = load_data()
yesterday_cards = data_now['flashcards'].get(yesterday_str, [])

health_recs = get_health_recommendations(weather_info, data_now['todos'], yesterday_cards)
print('\n👉 Now run Cell 8 — Your Morning Brief Dashboard!')

🤖 Calling HuggingFace AI... (may take 15–45 seconds on first call)
   Model is loading if this is your first call today — please wait.
   ⚠️  Attempt 1 failed: 410 Client Error: Gone for url: https://api-inference.huggingface.co/models/HuggingFaceH4/zephyr-7b-beta
   ⚠️  Attempt 2 failed: 410 Client Error: Gone for url: https://api-inference.huggingface.co/models/HuggingFaceH4/zephyr-7b-beta
   ⚠️  Attempt 3 failed: 410 Client Error: Gone for url: https://api-inference.huggingface.co/models/HuggingFaceH4/zephyr-7b-beta

⚠️  HuggingFace API unavailable. Using smart fallback recommendations.

🧬 SMART HEALTH RECOMMENDATIONS FOR TODAY
1. 🚶 UV index is low — great day for an outdoor walk or light exercise.
2. 💧 High humidity today — drink an extra 500ml of water and stay in shade.
3. 🌿 Great temperature for a 20-minute morning walk to boost your metabolism.
4. 🧠 Follow the 20-20-20 rule: every 20 minutes look 20 feet away for 20 seconds.
5. 📚 No flashcards yesterday — start small: add 5 car

In [ ]:
# ============================================================
# CELL 7-A  —  INSTALL gTTS  (run once per session)
# ============================================================

!pip install gtts -q

from gtts import gTTS
from IPython.display import display, Audio, HTML
import os, re, textwrap

# ── Voice language setting ───────────────────────────────────
# Full list: https://gtts.readthedocs.io/en/latest/module.html#languages-gtts-lang
VOICE_LANG  = 'en'     # 'en' = English (clear & natural)
VOICE_SLOW  = False    # True = slower, easier to follow
TTS_FILE    = '/content/neuro_brief_voice.mp3'

print('✅ gTTS installed and ready!')
print(f'   Language : {VOICE_LANG}')
print(f'   Slow mode: {VOICE_SLOW}')
print()
print('👉 Now run Cell 7-B to generate and play your health brief.')

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.2/98.2 kB 5.3 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
typer 0.24.1 requires click>=8.2.1, but you have click 8.1.8 which is incompatible.
✅ gTTS installed and ready!
   Language : en
   Slow mode: False

👉 Now run Cell 7-B to generate and play your health brief.


In [ ]:
# ============================================================
# CELL 7-B  —  SPEAK THE HEALTH BRIEF
# Generates MP3 with gTTS and plays it inline
# ============================================================

def clean_for_speech(text):
    """Strip emoji, markdown symbols, and extra spaces for clean TTS."""
    # Remove emoji (broad unicode range)
    text = re.sub(
        r'[\U0001F300-\U0001FAFF\U00002702-\U000027B0\U0001F000-\U0001F02F'
        r'\U0001F0A0-\U0001F0FF\U0001F100-\U0001F1FF\U0001F200-\U0001F2FF'
        r'\U0001F004\U0001F0CF\U0001F170-\U0001F171\U0001F17E-\U0001F17F'
        r'\u2600-\u26FF\u2700-\u27BF]+', '', text)
    # Remove markdown-style symbols
    text = re.sub(r'[\*\_\#\`\~\>\|]', '', text)
    # Normalise whitespace
    text = re.sub(r'\s+', ' ', text).strip()
    return text


def speak_health_brief_gtts(recs_text=None):
    """
    Convert the AI health brief to speech using gTTS and play it.
    Pass recs_text manually, or leave None to auto-use health_recs.
    """
    # ── 1. Get the text ──────────────────────────────────────
    try:
        raw = recs_text or health_recs or ''
    except NameError:
        raw = ''

    if not raw:
        print('⚠️  No health recommendations found.')
        print('   Please run Cell 7 first, then re-run this cell.')
        return

    # ── 2. Build the spoken script ───────────────────────────
    try:   name = YOUR_NAME
    except NameError: name = 'friend'

    try:   city = weather_info.get('city', 'your city')
    except NameError: city = 'your city'

    try:   temp = weather_info.get('temp_c', '?')
    except NameError: temp = '?'

    try:   cond = weather_info.get('desc', '')
    except NameError: cond = ''

    intro = (
        f"Good morning, {name}! "
        f"Welcome to your Neuro Band Morning Brief. "
        f"Today in {city}, it is {temp} degrees celsius with {cond}. "
        f"Here are your personalised health recommendations for today. "
    )
    outro = (
        " That's your health brief for today. "
        "Stay focused, drink plenty of water, and have a great day!"
    )

    full_script = intro + clean_for_speech(str(raw)) + outro

    print('📝 Script preview (first 200 chars):')
    print('  ', full_script[:200], '...')
    print()

    # ── 3. Generate MP3 ──────────────────────────────────────
    print('🎙️  Generating voice audio with gTTS...')
    try:
        tts = gTTS(text=full_script, lang=VOICE_LANG, slow=VOICE_SLOW)
        tts.save(TTS_FILE)
        size_kb = os.path.getsize(TTS_FILE) // 1024
        print(f'✅ Audio generated!  ({size_kb} KB)  →  {TTS_FILE}')
    except Exception as e:
        print(f'❌ gTTS error: {e}')
        print('   Check your internet connection and try again.')
        return

    # ── 4. Display the player card ───────────────────────────
    display(HTML("""
    <style>
      @import url('https://fonts.googleapis.com/css2?family=Syne:wght@700;800&family=DM+Sans:wght@400;500&display=swap');
    </style>
    <div style='
      background: linear-gradient(145deg,#0f0c29,#302b63);
      border: 1px solid rgba(255,255,255,.1);
      border-radius: 18px;
      padding: 22px 24px;
      max-width: 560px;
      box-shadow: 0 20px 60px rgba(0,0,0,.5);
      font-family: DM Sans, sans-serif;
      color: white;
    '>
      <div style='display:flex;align-items:center;gap:12px;margin-bottom:14px;'>
        <span style='font-size:2em;'>🔊</span>
        <div>
          <div style='font-family:Syne,sans-serif;font-weight:800;font-size:1.05em;
                      letter-spacing:2px;'>NEURO BAND VOICE</div>
          <div style='color:#7b8cde;font-size:.78em;'>Health Brief — spoken by gTTS</div>
        </div>
        <span style='margin-left:auto;background:rgba(46,213,115,.2);
                     color:#2ed573;padding:3px 10px;border-radius:20px;font-size:.75em;'>
          ✅ Audio Ready
        </span>
      </div>
      <div style='background:rgba(255,255,255,.05);border-radius:10px;
                  padding:10px 14px;margin-bottom:14px;
                  border-left:3px solid #667eea;font-size:.82em;
                  color:#c5cae9;line-height:1.6;'>
        Press ▶ on the audio player below to hear your brief.
      </div>
    </div>
    """))

    # ── 5. Play the audio inline ─────────────────────────────
    display(Audio(TTS_FILE, autoplay=True))

    print()
    print('🔊 Audio player shown above — press ▶ to play!')
    print('   Tip: If autoplay is blocked, just click the ▶ button manually.')


# ── Run it ────────────────────────────────────────────────────
speak_health_brief_gtts()
print('\n👉 Now run Cell 8 for the full dashboard with built-in voice button!')

📝 Script preview (first 200 chars):
   Good morning, Friend! Welcome to your Neuro Band Morning Brief. Today in Colombo, it is 27 degrees celsius with Partly cloudy. Here are your personalised health recommendations for today. 1. UV index  ...

🎙️  Generating voice audio with gTTS...
✅ Audio generated!  (439 KB)  →  /content/neuro_brief_voice.mp3



🔊 Audio player shown above — press ▶ to play!
   Tip: If autoplay is blocked, just click the ▶ button manually.

👉 Now run Cell 8 for the full dashboard with built-in voice button!


In [ ]:
# ============================================================
# CELL 8  —  MORNING BRIEF DASHBOARD  (with 🔊 Speak button)
# This replaces the old Cell 8 entirely
# ============================================================

import ipywidgets as widgets
from IPython.display import display, HTML, Audio, clear_output
from datetime import datetime, date, timedelta
import os, re

def generate_morning_brief_v2():
    data       = load_data()
    today_date = datetime.now().strftime('%A, %B %d, %Y')
    today_time = datetime.now().strftime('%I:%M %p')

    yest_cards = data['flashcards'].get(yesterday_str, [])
    pending    = [t for t in data['todos'] if not t['done']]
    done_tasks = [t for t in data['todos'] if  t['done']]
    yest_topics= list(set(c.get('topic','General') for c in yest_cards))

    w = weather_info

    # ── Weather grid ─────────────────────────────────────────
    weather_html = f"""
    <div style='display:grid;grid-template-columns:repeat(4,1fr);gap:8px;margin:10px 0;'>
      {''.join([
        f"<div style='background:rgba(255,255,255,.1);padding:10px;border-radius:8px;text-align:center;'>"
        f"<div style='font-size:1.5em'>{icon}</div>"
        f"<div style='font-size:1em;font-weight:700'>{val}</div>"
        f"<div style='opacity:.55;font-size:.72em'>{label}</div></div>"
        for icon, val, label in [
            ('🌡️', f"{w['temp_c']}°C", f"feels {w['feels_c']}°C"),
            ('💧', f"{w['humidity']}%", 'Humidity'),
            ('☀️', w['uv_index'],       'UV Index'),
            ('💨', f"{w['wind_kmph']} km/h", 'Wind'),
        ]
      ])}
    </div>
    <div style='text-align:center;opacity:.65;font-size:.82em;'>
      {w['desc']} &nbsp;|&nbsp; {w['min_c']}°C – {w['max_c']}°C today
    </div>"""

    # ── Flashcards ────────────────────────────────────────────
    if yest_cards:
        items = ''.join([
            f"<li style='margin:5px 0'>"
            f"<span style='background:rgba(102,126,234,.45);padding:1px 7px;"
            f"border-radius:10px;font-size:.74em;'>{c.get('topic','?')}</span> "
            f"{c['question'][:55]}{'…' if len(c['question'])>55 else ''}</li>"
            for c in yest_cards[:6]
        ])
        extra   = f"<li style='opacity:.45'>…and {len(yest_cards)-6} more</li>" if len(yest_cards)>6 else ""
        fc_html = f"<ul style='margin:6px 0;padding-left:18px;'>{items}{extra}</ul>"
    else:
        fc_html = "<p style='opacity:.45;margin:6px 0;'>No flashcards yesterday — add some in Cell 4!</p>"

    # ── To-do ─────────────────────────────────────────────────
    if pending:
        prio_order  = {'🔴 High':0,'🟡 Medium':1,'🟢 Low':2}
        badge_color = {'🔴 High':'#ff4757','🟡 Medium':'#ffa502','🟢 Low':'#2ed573'}
        sp = sorted(pending, key=lambda t: prio_order.get(t['priority'],1))
        todo_items = ''.join([
            f"<li style='margin:5px 0;'>"
            f"<span style='background:{badge_color.get(t['priority'],'#888')};"
            f"padding:1px 7px;border-radius:10px;font-size:.74em;color:#fff;'>"
            f"{t['priority'].split()[-1].upper()}</span> {t['task']}</li>"
            for t in sp[:8]
        ])
        todo_html = f"<ul style='margin:6px 0;padding-left:18px;'>{todo_items}</ul>"
    else:
        todo_html = "<p style='opacity:.45;margin:6px 0;'>All caught up! No pending tasks.</p>"

    # ── Health text ───────────────────────────────────────────
    health_html = str(health_recs).replace('\n','<br>') if health_recs else \
        'Run Cell 7 to generate recommendations.'

    # ── Render the dashboard ──────────────────────────────────
    display(HTML(f"""
    <style>
      @import url('https://fonts.googleapis.com/css2?family=Syne:wght@400;700;800&family=DM+Sans:wght@300;400;500&display=swap');
      .nb2 * {{ font-family:'DM Sans',sans-serif; color:white; box-sizing:border-box; }}
      .nb2 h1 {{ font-family:'Syne',sans-serif; }}
      .sec {{ background:rgba(255,255,255,.06); border-radius:14px;
               padding:14px 16px; margin-bottom:14px; }}
      .lbl {{ font-size:.73em; letter-spacing:2px; opacity:.5;
               font-family:'Syne',sans-serif; text-transform:uppercase; margin-bottom:6px; }}
    </style>

    <div class='nb2' style='
      background:linear-gradient(145deg,#0f0c29 0%,#302b63 55%,#1a1a2e 100%);
      padding:28px 24px; border-radius:22px; max-width:720px;
      border:1px solid rgba(255,255,255,.08);
      box-shadow:0 30px 80px rgba(0,0,0,.6);
    '>
      <div style='text-align:center;margin-bottom:22px;
                  padding-bottom:18px;border-bottom:1px solid rgba(255,255,255,.1);'>
        <div style='font-size:2.6em;margin-bottom:4px;'>🧠⚡</div>
        <h1 style='margin:0;font-size:1.9em;font-weight:800;letter-spacing:3px;'>NEURO BAND</h1>
        <p style='margin:3px 0 0;opacity:.4;font-size:.78em;letter-spacing:4px;'>MORNING BRIEF SYSTEM</p>
        <div style='margin-top:12px;display:flex;justify-content:center;gap:20px;
                    opacity:.65;font-size:.83em;flex-wrap:wrap;'>
          <span>👋 Good morning, {YOUR_NAME}!</span>
          <span>📅 {today_date}</span>
          <span>⏰ {today_time}</span>
        </div>
        <div style='margin-top:10px;display:flex;justify-content:center;gap:12px;flex-wrap:wrap;'>
          <span style='background:rgba(102,126,234,.3);padding:3px 12px;border-radius:20px;font-size:.78em;'>📚 {len(yest_cards)} cards yesterday</span>
          <span style='background:rgba(17,153,142,.3);padding:3px 12px;border-radius:20px;font-size:.78em;'>✅ {len(pending)} pending</span>
          <span style='background:rgba(255,71,87,.3);padding:3px 12px;border-radius:20px;font-size:.78em;'>🏆 {len(done_tasks)} done</span>
        </div>
      </div>

      <div class='sec'><div class='lbl'>🌤️ Weather — {w['city']}</div>{weather_html}</div>
      <div class='sec'><div class='lbl'>📚 Yesterday's Flashcards — {len(yest_cards)} | {', '.join(yest_topics) if yest_topics else 'No topics'}</div>{fc_html}</div>
      <div class='sec'><div class='lbl'>✅ To-Do — {len(pending)} pending / {len(done_tasks)} done</div>{todo_html}</div>

      <div style='background:rgba(102,126,234,.15);border:1px solid rgba(102,126,234,.35);
                  border-radius:14px;padding:14px 16px;margin-bottom:14px;'>
        <div class='lbl'>🧬 AI Health Brief — HuggingFace</div>
        <div style='line-height:1.75;font-size:.87em;opacity:.88;'>{health_html}</div>
      </div>

      <div style='text-align:center;opacity:.22;font-size:.68em;letter-spacing:2px;'>
        NEURO BAND MORNING BRIEF • {today_date}
      </div>
    </div>
    """))

    # ── Voice button (widget, always works) ──────────────────
    voice_out = widgets.Output()

    speak_btn = widgets.Button(
        description  = '🔊  Speak Health Brief',
        button_style = 'primary',
        layout       = widgets.Layout(width='220px', height='40px')
    )

    def on_speak(b):
        with voice_out:
            clear_output()
            print('🎙️  Generating audio...')
            try:
                raw = str(health_recs) if health_recs else 'No health data.'
                intro = (
                    f"Good morning, {YOUR_NAME}! Here is your health brief. "
                    f"Today in {w.get('city','your city')}, "
                    f"it is {w.get('temp_c','?')} degrees celsius with {w.get('desc','')}. "
                    f"Here are your health recommendations. "
                )
                outro = " Have a wonderful and productive day!"
                # Clean emoji from raw text
                clean = re.sub(
                    r'[\U0001F300-\U0001FAFF\u2600-\u26FF\u2700-\u27BF]+', '', raw)
                clean = re.sub(r'\s+', ' ', clean).strip()
                script = intro + clean + outro

                tts = gTTS(text=script, lang=VOICE_LANG, slow=VOICE_SLOW)
                tts.save(TTS_FILE)
                print('✅ Playing now — look for the audio player below!')
                display(Audio(TTS_FILE, autoplay=True))
            except Exception as e:
                print(f'❌ Error: {e}')

    speak_btn.on_click(on_speak)

    display(
        widgets.HTML(value="<div style='margin:10px 0 4px;font-family:sans-serif;"
                           "color:#888;font-size:.82em;'>🔊 Click to hear your health brief spoken aloud:</div>"),
        speak_btn,
        voice_out
    )

generate_morning_brief_v2()

HTML(value="<div style='margin:10px 0 4px;font-family:sans-serif;color:#888;font-size:.82em;'>🔊 Click to hear …

Button(button_style='primary', description='🔊  Speak Health Brief', layout=Layout(height='40px', width='220px'…

Output()

---
## 📌 Daily Usage Guide

| When | What to do |
|------|------------|
| **Every morning** | Run Cell 1 (install) → Cell 3 (data) → Cell 6 (weather) → Cell 7 (AI) → Cell 8 (brief) |
| **During study sessions** | Run Cell 4 to add flashcards as you study |
| **Planning your day** | Run Cell 5 to add/manage tasks |
| **End of day** | Run Cell 5 → mark tasks done → run Cell 8 to see updated brief |

### ⚡ Quick Re-run (after first setup)
Every new Colab session, just run **Cells 1 → 3 → 6 → 7 → 8** to get your Morning Brief instantly.

---
**Data persists** in `/content/neuro_band_data.json` during a session.  
**Tip:** Download this file to your Google Drive before closing Colab to save your data permanently.

In [ ]:
# ============================================================
# CELL 7-A — VOICE ENGINE SETUP
# Uses Web Speech API (FREE, built into Chrome/Edge)
# Run once per session, before Cell 7-B
# ============================================================

from IPython.display import display, HTML, Javascript
import json

# ─── Voice settings — tweak these to your preference ────────
VOICE_SETTINGS = {
    'rate'  : 0.92,   # Speed: 0.5 (slow) → 1.0 (normal) → 1.5 (fast)
    'pitch' : 1.05,   # Pitch: 0.5 (deep) → 1.0 (normal) → 2.0 (high)
    'volume': 1.0,    # Volume: 0.0 (mute) → 1.0 (max)
    # Voice name — best free voices by browser:
    # Chrome on Windows : 'Google US English'  or 'Microsoft David'
    # Chrome on Mac     : 'Samantha'           or 'Karen'
    # Chrome on Linux   : 'Google UK English Female'
    # Leave as '' to use browser default
    'voice_name': 'Google US English'
}

# ─── Inject the voice engine into the Colab browser ─────────
display(HTML(r"""
<div id='nb-voice-status'
     style='font-family:sans-serif;padding:10px 16px;border-radius:8px;
            background:#1a1a2e;color:#a0a8ff;font-size:.85em;display:inline-block;'>
  🔊 Voice engine initialising…
</div>
<script>
// ── Store settings globally in browser window ────────────────
window.nbVoiceSettings = {
  rate      : 0.92,
  pitch     : 1.05,
  volume    : 1.0,
  voiceName : 'Google US English'
};

// ── Core speak function ──────────────────────────────────────
window.nbSpeak = function(text, onDone) {
  if (!window.speechSynthesis) {
    alert('Speech not supported. Please use Chrome or Edge.');
    return;
  }
  window.speechSynthesis.cancel();

  // Clean the text: remove HTML tags and special chars
  var clean = text
    .replace(/<[^>]*>/g, ' ')
    .replace(/[🧬📚✅🌤️💧🧠☀️💨🌡️⚡🧠🔴🟡🟢🔊]/gu, '')
    .replace(/[\u{1F300}-\u{1FFFF}]/gu, '')
    .replace(/&nbsp;/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();

  var utter  = new SpeechSynthesisUtterance(clean);
  var cfg    = window.nbVoiceSettings;
  utter.rate   = cfg.rate;
  utter.pitch  = cfg.pitch;
  utter.volume = cfg.volume;

  // Try to find the requested voice
  var voices = window.speechSynthesis.getVoices();
  if (cfg.voiceName && voices.length > 0) {
    var match = voices.find(v => v.name.includes(cfg.voiceName));
    if (!match) match = voices.find(v => v.lang.startsWith('en'));
    if (match) utter.voice = match;
  }

  utter.onend   = function() { if (onDone) onDone(); };
  utter.onerror = function(e) { console.warn('TTS error:', e.error); };

  window.speechSynthesis.speak(utter);
};

// ── Pause / Resume / Stop ───────────────────────────────────
window.nbPause  = function() { window.speechSynthesis.pause();  };
window.nbResume = function() { window.speechSynthesis.resume(); };
window.nbStop   = function() { window.speechSynthesis.cancel(); };

// ── Wait for voices to load (Chrome loads them async) ───────
function initVoice() {
  var s = document.getElementById('nb-voice-status');
  if (window.speechSynthesis) {
    if (window.speechSynthesis.onvoiceschanged !== undefined) {
      window.speechSynthesis.onvoiceschanged = function() {
        var v = window.speechSynthesis.getVoices();
        s.innerHTML = '✅ Voice engine ready! Found ' + v.length + ' voice(s). Run Cell 7-B to speak.';
        s.style.color = '#2ed573';
      };
    }
    // Also trigger immediately in case voices are already loaded
    window.speechSynthesis.getVoices();
    s.innerHTML = '✅ Voice engine ready! Run Cell 7-B to speak the health brief.';
    s.style.color = '#2ed573';
  } else {
    s.innerHTML = '❌ Speech not supported. Please use Chrome or Edge browser.';
    s.style.color = '#ff4757';
  }
}
initVoice();
</script>
"""))

print('\n✅ Voice engine injected into browser!')
print('   Settings:', VOICE_SETTINGS)
print()
print('📌 TIPS:')
print('   • Use Chrome or Edge for best voice quality')
print('   • If you hear no sound, check your system volume')
print('   • Change VOICE_SETTINGS above and re-run to adjust')
print()
print('👉 Now run Cell 7-B to speak the AI health brief!')


✅ Voice engine injected into browser!
   Settings: {'rate': 0.92, 'pitch': 1.05, 'volume': 1.0, 'voice_name': 'Google US English'}

📌 TIPS:
   • Use Chrome or Edge for best voice quality
   • If you hear no sound, check your system volume
   • Change VOICE_SETTINGS above and re-run to adjust

👉 Now run Cell 7-B to speak the AI health brief!


In [ ]:
# ============================================================
# CELL 7-B — SPEAK THE AI HEALTH BRIEF
# Full interactive player with play/pause/stop/replay
# Run AFTER Cell 7 (AI Health Recommendations)
# ============================================================

import json

def speak_health_brief(recs_text=None):
    """
    Display an interactive voice player and speak the health brief.
    recs_text: pass your health_recs variable, or leave None to auto-use it.
    """
    # ── Get the text ─────────────────────────────────────────
    try:
        text = recs_text or health_recs or 'No health recommendations found. Please run Cell 7 first.'
    except NameError:
        text = 'No health recommendations found. Please run Cell 7 first.'

    # ── Build an intro ────────────────────────────────────────
    try:
        greeting = f'Good morning, {YOUR_NAME}! Here is your Neuro Band health brief for today.'
    except NameError:
        greeting = 'Good morning! Here is your Neuro Band health brief for today.'

    full_speech = greeting + ' ' + str(text)

    # ── Escape text for JS ────────────────────────────────────
    safe_text = json.dumps(full_speech)

    display(HTML(f"""
    <style>
      @import url('https://fonts.googleapis.com/css2?family=Syne:wght@700;800&family=DM+Sans:wght@300;400;500&display=swap');
      .vp * {{ font-family: 'DM Sans', sans-serif; box-sizing: border-box; }}
      .vp-btn {{
        border: none; cursor: pointer; border-radius: 50px;
        font-size: 1em; font-weight: 600; padding: 10px 22px;
        transition: all .2s; display: inline-flex;
        align-items: center; gap: 6px; outline: none;
      }}
      .vp-btn:hover {{ transform: translateY(-2px); filter: brightness(1.15); }}
      .vp-btn:active {{ transform: translateY(0); }}
      .vp-play  {{ background: linear-gradient(135deg,#667eea,#764ba2); color:#fff; }}
      .vp-pause {{ background: linear-gradient(135deg,#f093fb,#f5576c); color:#fff; }}
      .vp-stop  {{ background: rgba(255,255,255,.1); color:#ccc; border:1px solid rgba(255,255,255,.15); }}
      .vp-replay{{ background: linear-gradient(135deg,#11998e,#38ef7d); color:#fff; }}
      .vp-wave {{
        display: flex; align-items: center; gap: 3px;
        height: 24px; margin-left: 4px;
      }}
      .vp-bar {{
        width: 4px; border-radius: 2px;
        background: linear-gradient(to top, #667eea, #a8edea);
        animation: none;
      }}
      .vp-bar.active {{
        animation: wave .8s ease-in-out infinite;
      }}
      .vp-bar:nth-child(1){{ height:6px;  animation-delay:0s;    }}
      .vp-bar:nth-child(2){{ height:14px; animation-delay:.1s;   }}
      .vp-bar:nth-child(3){{ height:20px; animation-delay:.2s;   }}
      .vp-bar:nth-child(4){{ height:14px; animation-delay:.3s;   }}
      .vp-bar:nth-child(5){{ height:8px;  animation-delay:.15s;  }}
      .vp-bar:nth-child(6){{ height:18px; animation-delay:.25s;  }}
      .vp-bar:nth-child(7){{ height:12px; animation-delay:.05s;  }}
      @keyframes wave {{
        0%,100% {{ transform: scaleY(1);   }}
        50%      {{ transform: scaleY(2.5); }}
      }}
      .vp-speed {{
        background: rgba(255,255,255,.08);
        border: 1px solid rgba(255,255,255,.15);
        color: #ccc; border-radius: 8px;
        padding: 4px 10px; font-size: .8em; cursor: pointer;
      }}
    </style>

    <div class='vp' style='
      background: linear-gradient(145deg,#0f0c29,#302b63);
      border: 1px solid rgba(255,255,255,.08);
      border-radius: 18px;
      padding: 22px 24px;
      max-width: 580px;
      box-shadow: 0 20px 60px rgba(0,0,0,.5);
    '>

      <!-- Title -->
      <div style='display:flex;align-items:center;gap:10px;margin-bottom:16px;'>
        <span style='font-size:1.8em;'>🔊</span>
        <div>
          <div style='font-family:Syne,sans-serif;font-weight:800;font-size:1.1em;
                      color:#fff;letter-spacing:1px;'>HEALTH BRIEF PLAYER</div>
          <div style='color:#7b8cde;font-size:.75em;'>Neuro Band Voice System</div>
        </div>
        <!-- Animated wave bars -->
        <div class='vp-wave' style='margin-left:auto;'>
          <div class='vp-bar' id='b1'></div>
          <div class='vp-bar' id='b2'></div>
          <div class='vp-bar' id='b3'></div>
          <div class='vp-bar' id='b4'></div>
          <div class='vp-bar' id='b5'></div>
          <div class='vp-bar' id='b6'></div>
          <div class='vp-bar' id='b7'></div>
        </div>
      </div>

      <!-- Status bar -->
      <div id='vp-status' style='
        background: rgba(255,255,255,.05);
        border-radius: 8px; padding: 8px 12px;
        color: #a0a8ff; font-size: .8em; margin-bottom: 14px;
        border-left: 3px solid #667eea;
      '>⏸ Ready — click Play to hear your health brief</div>

      <!-- Text preview -->
      <div id='vp-text' style='
        background: rgba(255,255,255,.04);
        border-radius: 10px; padding: 10px 14px;
        color: #c5cae9; font-size: .82em; line-height: 1.7;
        max-height: 100px; overflow-y: auto;
        margin-bottom: 16px;
        border: 1px solid rgba(255,255,255,.06);
      '>{str(text).replace(chr(10), '<br>')}</div>

      <!-- Controls -->
      <div style='display:flex;gap:8px;flex-wrap:wrap;align-items:center;margin-bottom:14px;'>
        <button class='vp-btn vp-play'  id='vp-play'  onclick='vpPlay()'>▶ Play</button>
        <button class='vp-btn vp-pause' id='vp-pause' onclick='vpPause()' style='display:none;'>⏸ Pause</button>
        <button class='vp-btn vp-play'  id='vp-resume'onclick='vpResume()' style='display:none;'>▶ Resume</button>
        <button class='vp-btn vp-stop'  id='vp-stop'  onclick='vpStop()'>■ Stop</button>
        <button class='vp-btn vp-replay'onclick='vpReplay()'>↺ Replay</button>
      </div>

      <!-- Speed & Pitch sliders -->
      <div style='display:grid;grid-template-columns:1fr 1fr;gap:12px;'>
        <div>
          <label style='color:#7b8cde;font-size:.75em;letter-spacing:1px;'>SPEED
            <span id='rate-val' style='color:#fff;margin-left:6px;'>0.92×</span>
          </label>
          <input type='range' id='rate-slider' min='0.5' max='1.6' step='0.05' value='0.92'
                 oninput="window.nbVoiceSettings.rate=parseFloat(this.value);
                          document.getElementById('rate-val').textContent=parseFloat(this.value).toFixed(2)+'×';"
                 style='width:100%;accent-color:#667eea;cursor:pointer;'>
        </div>
        <div>
          <label style='color:#7b8cde;font-size:.75em;letter-spacing:1px;'>PITCH
            <span id='pitch-val' style='color:#fff;margin-left:6px;'>1.05</span>
          </label>
          <input type='range' id='pitch-slider' min='0.5' max='2.0' step='0.05' value='1.05'
                 oninput="window.nbVoiceSettings.pitch=parseFloat(this.value);
                          document.getElementById('pitch-val').textContent=parseFloat(this.value).toFixed(2);"
                 style='width:100%;accent-color:#f093fb;cursor:pointer;'>
        </div>
      </div>

    </div>

    <script>
    var VP_TEXT = {safe_text};

    function setWave(active) {{
      ['b1','b2','b3','b4','b5','b6','b7'].forEach(function(id) {{
        var el = document.getElementById(id);
        if (el) active ? el.classList.add('active') : el.classList.remove('active');
      }});
    }}

    function setStatus(msg, color) {{
      var s = document.getElementById('vp-status');
      if (s) {{ s.textContent = msg; s.style.color = color || '#a0a8ff'; }}
    }}

    function showBtn(play, pause, resume) {{
      document.getElementById('vp-play').style.display   = play   ? 'inline-flex' : 'none';
      document.getElementById('vp-pause').style.display  = pause  ? 'inline-flex' : 'none';
      document.getElementById('vp-resume').style.display = resume ? 'inline-flex' : 'none';
    }}

    function vpPlay() {{
      showBtn(false, true, false);
      setStatus('🔊 Speaking your health brief…', '#2ed573');
      setWave(true);
      window.nbSpeak(VP_TEXT, function() {{
        setStatus('✅ Done! Click Replay to hear again.', '#a0a8ff');
        setWave(false);
        showBtn(true, false, false);
      }});
    }}

    function vpPause() {{
      window.nbPause();
      showBtn(false, false, true);
      setStatus('⏸ Paused.', '#ffa502');
      setWave(false);
    }}

    function vpResume() {{
      window.nbResume();
      showBtn(false, true, false);
      setStatus('🔊 Resuming…', '#2ed573');
      setWave(true);
    }}

    function vpStop() {{
      window.nbStop();
      showBtn(true, false, false);
      setStatus('■ Stopped.', '#ff4757');
      setWave(false);
    }}

    function vpReplay() {{
      vpStop();
      setTimeout(vpPlay, 300);
    }}
    </script>
    """))
    print('\n✅ Voice player displayed above!')
    print('   Click ▶ Play to hear your health brief spoken aloud.')
    print('   Adjust Speed and Pitch sliders for your preferred voice style.')


# ── Launch the player ─────────────────────────────────────────
speak_health_brief()
print('\n👉 After listening, run Cell 8 for the full Morning Brief Dashboard!')


✅ Voice player displayed above!
   Click ▶ Play to hear your health brief spoken aloud.
   Adjust Speed and Pitch sliders for your preferred voice style.

👉 After listening, run Cell 8 for the full Morning Brief Dashboard!


Voice Quick Reference
Cell	Purpose	When to run
7-A	Load voice engine into browser	Once per session, before 7-B
7-B	Standalone voice player	After Cell 7 (AI health)
8	Full dashboard with built-in Speak button	Any time
🎛️ Adjust Your Voice
In Cell 7-A, change VOICE_SETTINGS:

VOICE_SETTINGS = {
    'rate'      : 0.85,              # Slower and clearer
    'pitch'     : 1.0,               # Natural pitch
    'volume'    : 1.0,               # Max volume
    'voice_name': 'Google UK English Female'  # British female voice
}
🌍 Best Free Voice Names by Browser
Browser	Best voice name
Chrome Windows	Google US English
Chrome Mac	Samantha
Chrome Linux	Google UK English Female
Edge	Microsoft Aria Online
✅ No API key, no cost, no limit — uses your browser's built-in speech engine.

